indico/modules/categories/templates/display/base.html

```diff
@@ -107,19 +107,11 @@ <h1 class="category-title">
         </div>
     </div>
 </div>
-<script>
-    $('#navigate-button').on('click', function(evt) {
-        $('<div>').categorynavigator({
-            openInDialog: true,
-            category: {{ category.id }},
-            actionButtonText: $T.gettext('Navigate to'),
-            dialogTitle: $T.gettext('Navigate to a category'),
-            onAction: function(category) {
-                window.location.href = build_url(Indico.Urls.Categories.display, {category_id: category.id});
-            }
-        });
-    });
-</script>
+
+<ind-category-navigator triggeredby="navigate-button"
```

Comment:
This is just one place where the category picker dialog is used.

We currently use the same dialog when creating events or moving events/categories. Since we don't want to keep two implementations of the same thing around, it would be good to adapt those parts as well in a future PR after this one is merged.

Reply:
I think it would be good to have one of you do it, with me as an assistant. Over time there's probably going to be more of this type of code, so it's good to become familiar. You know, while I'm still around. 😂

Reply:
The problem is that we're quite busy atm, so I don't think "adding big functionality to this component" is a thing one of us really has the time for right now...

I'd be happy to do the integration in those places (create event, move event, move category, badge template cloning), as long as the features needed for those particular operations are already present in the new category picker:

```js
{
  actionOn: {
    // UNUSED (i think)
    categoriesWithoutEventProposalRights: {
      message: $T.gettext('Not possible for categories where you cannot propose events'),
    },
    // REQUIRED - when moving events
    categoriesWithoutEventProposalOrCreationRights: {
      message: $T.gettext('Not possible for categories where you cannot propose/create events'),
    },
    // REQUIRED - when creating events
    categoriesWithoutEventCreationRights: {
      message: $T.gettext('Not possible for categories where you cannot create events'),
    },
    // REQUIRED - when cloning badge templates or moving categories [the latter was not being checked, i just fixed that in master, commit cac224f8fa] 
    categoriesWithoutCategoryManagementRights: {
      message: $T.gettext('Not possible for categories where you are not a manager'),
    },
    // REQUIRED - when moving categories
    categoriesDescendingFrom: {
      ids: [],
      // The closest parent category title will be used if more than one parent is present in ids
      message: $T.gettext('Not possible for categories descending from category "{0}"'),
    },
    // REQUIRED - when moving categories
    categories: {
      disabled: false,
      message: $T.gettext('Not possible for this category'),
      ids: [],
      // Expects an Array of {ids: [...], message: '...'} for more specific messages
      groups: [],
    },
  },
}
```

Reply:
It's not really "big functionality". All you customize is the button that triggers the action, which is supplied separately as a prop. But sure, no problem. I'll integrate it when I get around to those pages.

Reply:
But your new react picker does not have the logic for all those criteria yet, does it?

As I said, I'm fine integrating it on the pages once the logic is there ;)

(I want to avoid a long period of time of having two category pickers in the codebase in parallel)

Reply:
> But your new react picker does not have the logic for all those criteria yet, does it?

What logic are we talking about? Maybe we can hop on a call during the week, so you (or someone else on your team) can give me a quick summary?

Reply:
When instantiating the old jquery category picker you can pass some config arguments (`actionOn: {...}`) specifying when the action button (currently that's always your `NavigateTo` component) should be disabled (right now the message why it's disabled goes into a tooltip):

![image](https://github.com/indico/indico/assets/179599/694cc5b2-b165-4a86-aed4-1f35373386af)

For example when moving categories we pass this config:

```js
      actionOn: {
        categoriesWithoutCategoryManagementRights: {
          disabled: true,
        },
        categoriesDescendingFrom: {
          disabled: true,
          ids,
        },
        categories: {
          disabled: true,
          groups: [
            {
              ids: [sourceId],
              message: $T.ngettext(
                'The category is already here',
                'The selected categories are already here',
                ids.length
              ),
            },
            {
              ids,
              message: $T.ngettext(
                'This is the category you are trying to move',
                'This is one of the categories you are trying to move',
                ids.length
              ),
            },
          ],
        },
      },
```

Reply:
Thanks, I'll look into it.

Reply:
@ThiefMaster I've looked at this feature. 

From what I was able to gather by reading the code and trying the feature in the management section, it looks like the only thing this feature affects is:

1. Whether the button is disabled
2. The tooltip that is shown when the button is disabled

If these assumptions are correct, the good news is that the new implementation of the category navigator does not need this feature at all. It no longer deals with whether the button is disabled or not. In fact, it has no awareness of how the button is implemented or even whether it's a button. 

We pass it a component that will be rendered through the `actionView` prop. The component receives a category as its only prop, and is free to do whatever it wants. It can disable itself, change the tooltip, change the behavior, appearance, etc. Compared to the current implementation where the responsibility for every use case ever encountered in the app is addressed by the navigator component itself, this should greatly simplify the code.

If this explanation is insufficient, I think it's best that we sit down so I have a chance to walk you (or someone on your team) through the integration of at least one example, and then we can address any oversights and glitches on the spot.

Reply:
Ah, that sounds somewhat easy to adapt. The only problem I have right now is calling this in a purely imperative way from JS code where no DOM element is used at all (currently we trigger this from a click handler on a button)

Ideally I'd like to use it like this (e.g. inside `_moveCategories` when selecting the target category):

```js
const target = await injectModal(resolve => (
  <CategoryNavigator categoryId={sourceId} onClose={() => resolve(null)} onAction={cat => resolve(cat)} />
));
if (target) {
  console.log('selected', target);
} else {
  console.log('dialog closed w/o clicking the action button');
}
```

However, right now it seems to rely on creating the `ind-category-nav` element and triggering an event on `window` to open it (or instantiating one of them for each item so it can bind its event on the trigger).

I guess technically we could use one `ind-category-nav` for each move button, but then we'd still need to bind a JS event when clicking the action button (since it's not just navigating to a new URL when clicking those buttons but rather sending a request in the background). To me it feels cleaner if we could use `injectModel` (which mounts a React component and provides a promise to be resolved using a callback passed to the component): That way we have one place (the click handler for the move button) that first opens the picker, and then does something with its result.

Reply:
Actually, I forgot I changed the navigator to receive custom events to address that scenario.

So basically, it works like this now:

```
window.dispatchEvent(new CustomEvent('open-category-navigator', {detail: NavigateTo}));
```

The `detail` property is a React component that is used as a button.

You don't need to inject a modal. Render the navigator somewhere on the page, and it will set up a modal that will be open when the custom event is received.

Reply:
but we may have differently configured navigators on the same page (move category + move event for example), with different actions on button click

Reply:
the problem i see with the custom events is that it's very loosely coupled. while it should never happen (since it's shown in a modal after all), one could trigger the click event for two of the move buttons, and then have no way to know for which operation you have the category picker open...

Reply:
> What do you mean by "configuration" in this case?

- I move an event: Any other category is a valid target
- I move a subcategory: Any category that is a child of the category I'm moving is NOT a valid target

see the double-arrow icons on this page (category settings -> content). they all open a category picker:

![image](https://github.com/indico/indico/assets/179599/840eef03-c3f7-4be2-b541-cae893147ed8)

> That sounds like a bug. Why would you be able to move two categories at once?

you are right opening two modals at once would be a bug and is not going to happen unless someone plays around w/ dev tools. I just meant it as an example on why having to rely on global state is a bit ugly. because with being purely event-based I need to:

- store the relevant information, such as which category/event i'm trying to move, when clicking the "move" button next to it
- somehow provide the "config" (as explained above) to that category picker
- when receiving the event that a category has been selected, I need to use the data stored earlier in order to execute the move

at least to me this feels much more complicated than having a single event handler for the click event on the move button - in which I now have all the relevant data in scope, in there I could then simply await the picker selection and then perform the http request to execute the operation - and all the initial data is still in scope since it's all within a single (async) function.

Reply:
> Any other category is a valid target

What should this change about the category navigator?

> store the relevant information, such as which category/event i'm trying to move, when clicking the "move" button next to it

I don't think you need to do that. You can create a component factory function that will bind such information to the button component being passed to the navigator.

```
function createMoveButton(sourceId) {
  return ({category}) => {
    ....

    const handleClick = () => {
      // Here you have access to both sourceId and category
    }

    return (
      <button>....</button>
    )
  }
}

// ....

window.dispatchEvent(new CustomEvent('open-category-navigator', {detail: createMoveButton(123)}))
```

Reply:
Yeah, your example looks like it can do the job. I'll give it a try!